CFileType.js

```diff
@@ -81,70 +82,76 @@ CFileType.prototype.write = function(translations, locales) {
+                            baseTranslation = translated.getTarget();
+                        }
+                    }.bind(this));
+                }
+    
+                db.getResourceByCleanHashKey(res.cleanHashKeyForTranslation(locale), function(err, translated) {
+                    var r = translated;
+                    if (!translated || ( this.API.utils.cleanString(res.getSource()) !== this.API.utils.cleanString(r.getSource()) &&
+                        this.API.utils.cleanString(res.getSource()) !== this.API.utils.cleanString(r.getKey()))) {
+                        if (r) {
+                            this.logger.trace("extracted   source: " + this.API.utils.cleanString(res.getSource()));
+                            this.logger.trace("translation source: " + this.API.utils.cleanString(r.getSource()));
+                        }
+                        var note = r && 'The source string has changed. Please update the translation to match if necessary. Previous source: "' + r.getSource() + '"';
+                        var newres = res.clone();
+                        newres.setTargetLocale(locale);
+                        newres.setTarget((r && r.getTarget()) || res.getSource());
+                        newres.setState("new");
+                        newres.setComment(note);
+    
+                        this.newres.add(newres);
+    
+                        this.logger.trace("No translation for " + res.reskey + " to " + locale);
                     } else {
-                        this.logger.trace("Same translation as base translation for " + res.reskey + " to " + locale);
+                        if (res.reskey != r.reskey) {
+                            // if reskeys don't match, we matched on cleaned string.
+                            //so we need to overwrite reskey of the translated resource to match
+                            r = r.clone();
+                            r.reskey = res.reskey;
+                        }
+    
+                        if (baseTranslation != r.getTarget()) {
+                            file = resFileType.getResourceFile(locale);
+                            file.addResource(r);
+                            this.logger.trace("Added " + r.reskey + " to " + file.pathName);
+                        } else {
+                            this.logger.trace("Same translation as base translation for " + res.reskey + " to " + locale);
+                        }
                     }
-                }
+                }.bind(this));
             }.bind(this));
+        }
+    
+        resources = this.pseudo.getAll().filter(function(resource) {
+            return resource.datatype === this.datatype;
         }.bind(this));
     }
-
-    resources = this.pseudo.getAll().filter(function(resource) {
-        return resource.datatype === this.datatype;
-    }.bind(this));
+    else {
+        // generate mode
+        resources = this.project.db.ts.resources;
```

Comment:
Plugins should not be looking into the depths of private members because if any of the implementation details change from a refactor, then this will break. The plugins should not know how it all works inside. There should be a new public method getTranslations() on the Project class to get all of the translations, which in turn would call this.db.getBy() to retrieve them instead of looking into the private members of the LocalRepository class. The project is already exposed to the plugins, so they would be able to call that new public method. This would require a new version of the loctool to be published first.

Reply:
This applies to all of the PRs of this same type

Reply:
I see. Same changes is already applied to webos-javascript plugin. so I thought it's going to be ok to apply here too. I'll re-work following your suggestion.